kf_lib_data_ingest/validation/data_validator.py

```diff
@@ -0,0 +1,435 @@
+        # the hierarchy gaps test.)
+        for a, b in edges:
+            graph.add_edge(a, b)
+
+        logger.info(f"Final edge count: {len(graph.edges())}")
+        return graph, counts
+
+    def _get_node_counts(self, graph):
+        """Count the number of unique instances of each type of value.
+
+        :param graph: a graph representation of the data
+        :return: dict of type keys and count values
+        """
+        return {c: len(graph.groups[c]) for c in self.HIERARCHY_ORDER}
+
+    def _validate_graph_relationships(self, graph, dict_of_dataframes):
+        """Perform tests on the graph to validate which of the hierarchy
+        rules have been broken by the study data.
+
+        :param graph: a graph representation of the data
+        :param dict_of_dataframes: dict with filename keys and dataframe values
+        :yield: test result dicts
+        """
+        assert isinstance(graph, Graph)
+        assert len(graph.nodes()) > 0
+
+        logger.info("Validating graph relationships...")
+
+        reverse_graph = Graph()
+        for a, b, _ in graph.edges():
+            reverse_graph.add_edge(b, a)
+
+        membership_lookup = {
+            f: {col: set(df[col].values) for col in df.columns}
+            for f, df in dict_of_dataframes.items()
+        }
+
+        def find_in_files(node):
+            """Find which files a given node came from.
+
+            :param node: a node in the graph representing a data value
+            :return: a list of files
+            """
+            return [
+                f
+                for f, ml in membership_lookup.items()
+                if (node[0] in ml) and (node[1] in ml[node[0]])
+            ]
+
+        def cardinality(typeA, typeB, relation):
+            """Tests cardinality of connections between typeA and typeB.
+
+            :param typeA: a type node (nodes are inserted in the graph for value
+                types as well as values)
+            :param typeB: another type node
+            :param relation: one of the tuples from relations.RELATIONS
+            :return: a test result dict
+            """
+            desc, func = relation
+            logger.info(f"Testing {typeA} links to {desc} {typeB}...")
+            errors = deque()
+            A_nodes = graph.groups[typeA]
+            if typeB in self.HIERARCHY_DIRECT[typeA]:
+                g = graph
+            else:
+                g = reverse_graph
+            for n in A_nodes:
+                links = [c for c in g.nodes(n) if c[0] == typeB]
+                if not func(len(links)):
+                    locs = {m: find_in_files(m) for m in ([n] + links)}
+                    errors.append({"from": n, "to": links, "locations": locs})
+
+            desc = f"Each {typeA} links to {desc} {typeB}"
+            return self._format_result(
+                "relationship", desc, bool(A_nodes), errors, typeA, typeB
```

Comment:
This is logic is currently reporting false positives when it should report that the test did not run. I think its because you need to check whether both input columns exist for the relationship test, not just the left side (e.g. `A_nodes`)

Reply:
> it should report that the test did not run.

I think this is wrong. If the rule is that every A needs a B, and you have As but no Bs, then that's a failure.

Reply:
> You can see an example of this if you run the validator on tests/data/validation/DATASET1

Are you sure it's DATASET1?